File: src-tauri/src/handler/board.rs
use crate::db::board::{
    delete_board_card_from_db, delete_board_column_from_db, get_board_data_from_db,
    insert_board_card_into_db, insert_board_column_into_db, update_board_card_in_db,
    update_board_column_in_db,
};
use crate::models::board::{BoardCard, BoardColumn, BoardData, NewBoardCard, NewBoardColumn};
use crate::utils::db::AppDb;
use crate::utils::errors::DbResult;

#[tauri::command]
pub fn create_board_column(db: tauri::State<AppDb>, data: NewBoardColumn) -> DbResult<BoardColumn> {
    insert_board_column_into_db(&db, data)
}

#[tauri::command]
pub fn update_board_column(
    db: tauri::State<AppDb>,
    id: i64,
    title: String,
    order_index: i64,
    color: String,
) -> DbResult<()> {
    update_board_column_in_db(&db, id, title, order_index, color)
}

#[tauri::command]
pub fn delete_board_column(db: tauri::State<AppDb>, id: i64) -> DbResult<()> {
    delete_board_column_from_db(&db, id)
}

#[tauri::command]
pub fn create_board_card(db: tauri::State<AppDb>, data: NewBoardCard) -> DbResult<BoardCard> {
    insert_board_card_into_db(&db, data)
}

#[tauri::command]
pub fn update_board_card(
    db: tauri::State<AppDb>,
    id: i64,
    title: String,
    description: Option<String>,
    order_index: i64,
    column_id: i64,
) -> DbResult<()> {
    update_board_card_in_db(&db, id, title, description, order_index, column_id)
}

#[tauri::command]
pub fn delete_board_card(db: tauri::State<AppDb>, id: i64) -> DbResult<()> {
    delete_board_card_from_db(&db, id)
}

#[tauri::command]
pub fn get_board_data(db: tauri::State<AppDb>, nestling_id: i64) -> DbResult<BoardData> {
    get_board_data_from_db(&db, nestling_id)
}

File: src/lib/api/board.ts
import { invoke } from "@tauri-apps/api/core";
import {
  NewBoardColumn,
  BoardColumn,
  NewBoardCard,
  BoardCard,
  BoardData,
} from "../types/board";

export async function createBoardColumn(data: NewBoardColumn) {
  return await invoke<BoardColumn>("create_board_column", { data });
}

export async function updateBoardColumn({
  id,
  title,
  orderIndex,
  color,
}: {
  id: number;
  title: string;
  orderIndex: number;
  color: string;
}) {
  await invoke<void>("update_board_column", {
    id,
    title,
    orderIndex,
    color,
  });
}

export async function deleteBoardColumn(id: number) {
  await invoke<void>("delete_board_column", { id });
}

export async function createBoardCard(data: NewBoardCard) {
  return await invoke<BoardCard>("create_board_card", { data });
}

export async function updateBoardCard({
  id,
  title,
  description,
  orderIndex,
  columnId,
}: {
  id: number;
  title: string;
  description: string | null;
  orderIndex: number;
  columnId: number;
}) {
  await invoke<void>("update_board_card", {
    id,
    title,
    description,
    orderIndex,
    columnId,
  });
}
export async function deleteBoardCard(id: number) {
  await invoke<void>("delete_board_card", { id });
}

export async function getBoard(nestlingId: number) {
  return await invoke<BoardData>("get_board_data", { nestlingId });
}
